File: os/lab_08/01_parallel_sum_cpp/sum_threads.cpp
#include <iostream>
#include <vector>
#include <thread>
#include <mutex>
#include <random>
int main() {
    std::size_t n;
    if (!(std::cin >> n) || n < 16) return 0;
    std::vector<int> a(n);
    std::mt19937 rng(42);
    std::uniform_int_distribution<int> dist(0, 999);
    for (auto &v : a) v = dist(rng);
    std::size_t base= n/4;
    std::size_t rem  = n%4;
    long long sum = 0;
    std::mutex m;
    auto worker = [&](std::size_t l, std::size_t r) {
        long long local = 0;
        for (std::size_t i = l; i < r; ++i) local += a[i];
        std::lock_guard<std::mutex> g(m);
        sum += local;
    };
    std::vector<std::thread> th;
    std::size_t pos = 0;
    for (int i = 0; i < 4; ++i) {
        std::size_t len = base + (i < static_cast<int>(rem) ? 1u : 0u);
        th.emplace_back(worker, pos, pos + len);
        pos += len;
    }
    for (auto &t : th) t.join();
    std::cout << "Array:";
    for (int v : a) std::cout << ' ' << v;
    std::cout << '\n';
    std::cout << "Total sum: " << sum << '\n';
    return 0;
}

File: os/lab_08/01_parallel_sum_py/sum_threads.py
import sys, random, threading
line = sys.stdin.readline()
if not line:
    sys.exit(0)
n =int(line.strip())
if n<16:
    sys.exit(0)
random.seed(42)
a =[random.randint(0, 999) for _ in range(n)]
totl = 0
lock = threading.Lock()
def worker(l, r):
    local = sum(a[l:r])
    with lock:
        global totl
        totl += local
base, rem = divmod(n, 4)
threads = []
pos = 0
for i in range(4):
    length = base + (1 if i < rem else 0)
    t = threading.Thread(target=worker, args=(pos, pos + length))
    t.start()
    threads.append(t)
    pos += length

for t in threads:
    t.join()

print("Array:", *a)
print("Total sum:", totl)
